### PR TITLE
Possible refactoring proposal

### PR DESCRIPTION
**Pull request by mkohlmyr**
Feels like our example project is a little bit out of date. 

This is a refactor which uses a few newer react concepts, one or two es6 & typescript features and cleans up some of the ambiguous naming choices. 

Shouldn't be very hard for anyone who is already using the exercise to get up to date with this version. It is testing all the same areas of the code, it just houses them in slightly different places.

Feedback welcome.

### CODE REVIEW

File: src/components/Game.tsx [hunk 1]
```diff
@@ -1,44 +1,26 @@
-import * as React from "react";
-import { play, GESTURES, Gesture, Result } from "../game/game";
+import React, { useEffect, useState } from "react";
```
Comment (lucyewright) [inline src/components/Game.tsx]: can be: 
`import React, { FunctionComponent, useEffect, useState } from "react";`

File: src/components/Game.tsx [hunk 2]
```diff
@@ -1,44 +1,26 @@
...
-  winner: Result;
-  hasPlayed: boolean;
-};
-
-export default class Game extends React.Component<{}, State> {
-  state = {
-    winner: "Draw" as Result,
-    hasPlayed: false
-  };
+export const Game: React.FunctionComponent = ({ children }) => {
```
Comment (lucyewright) [inline src/components/Game.tsx:5]: can be: 
`export const Game: FunctionComponent<Props> = ({ children }) => {`

File: src/components/Game.tsx [hunk 3]
```diff
@@ -1,44 +1,26 @@
...
+  const [outcome, setOutcome] = useState(Outcome.None);
+  const [playerGesture, setPlayerGesture] = useState<Gesture | null>(null);
 
-  play(playerGesture: Gesture) {
-    const winner = play(playerGesture);
+  useEffect(() => {
+    if (playerGesture !== null) {
+      setOutcome(determineOutcome(playerGesture));
+    }
```
Comment (lucyewright) [inline src/components/Game.tsx]: does this need an else?

File: src/components/Game.tsx [hunk 4]
```diff
@@ -1,44 +1,26 @@
...
+  }, [playerGesture])
 
-    this.setState({ winner, hasPlayed: true });
-  }
+  const context = {
+    outcome,
+    playerGesture,
+    setPlayerGesture,
```
Comment (lucyewright) [inline src/components/Game.tsx]: why does the `GameProvider` need the `setPlayerGesture` function? I think I've not followed the logic correctly here?

File: src/components/Game.tsx (hunk 3 above)
Comment (clarencedglee) [reply src/components/Game.tsx]: Seems ok without the `else`, it can just return without doing anything.

File: src/components/Game.tsx (hunk 4 above)
Comment (clarencedglee) [reply src/components/Game.tsx]: This makes `setPlayerGesture` available to components lower down in the tree, e.g. in [GameControls](https://github.com/ovotech/react-rps-boilerplate/pull/9/files#diff-ac6b9ed17744b825a23e1d3b96e9528fdfae927f7f9266c360b9e7e2284ae0eeR27)

**Review by clarencedglee — APPROVED** [review]
This looks good. Always good to have a few visual emojis.

Also like that logic is just one function rather than split into two, so it's less leading.

Moving away from class components is very welcome.

**Comment by mkohlmyr** [comment]
Further simplified the context / state. 

Added two additional locations where the candidate is able to make some small choices.

@clarencedglee @lucyewright further feedback welcome 👍

**Review by lucyewright — APPROVED** [review]

**Comment by joepurnell1** [comment]
I have three thoughts on this

1. Not all of our codebases are using modern practices and so I think it's important to older implementation such as classes. This shows if the engineer understands the development react has taken and if they're able to adjust for some slightly unexpected implementations.

I've had interviews where candidates were strong with newer concepts but as soon as they had to work with classes they fell apart and it really highlighted their level.

2. I think there are too many TODOs here, it makes the exercise seem more like a box checking exercise rather than a evaluation on whether they can take in, understand, and move forward within a codebase.

3. Simplifying a lot of the typing logic to me takes away from allowing the candidate to highlight and talk on/improve which helps again to show differing levels in candidates.

TLDR, I think there are some minor improvements to make the exercise smoother which I agree on but it isn't a difficult test in the first place and I feel like handing over a perfect project to work off will make it harder for us to determine an accurate level of candidates

**Comment by mkohlmyr** [comment]
> I have three thoughts on this
> 
> 1. Not all of our codebases are using modern practices and so I think it's important to older implementation such as classes. This shows if the engineer understands the development react has taken and if they're able to adjust for some slightly unexpected implementations.
> 
> I've had interviews where candidates were strong with newer concepts but as soon as they had to work with classes they fell apart and it really highlighted their level.
> 
> 1. I think there are too many TODOs here, it makes the exercise seem more like a box checking exercise rather than a evaluation on whether they can take in, understand, and move forward within a codebase.
> 2. Simplifying a lot of the typing logic to me takes away from allowing the candidate to highlight and talk on/improve which helps again to show differing levels in candidates.
> 
> TLDR, I think there are some minor improvements to make the exercise smoother which I agree on but it isn't a difficult test in the first place and I feel like handing over a perfect project to work off will make it harder for us to determine an accurate level of candidates


I can happily remove the TODO comments to leave things more open ended, that's entirely fair. For the third point is there anything specific you have seen candidates do where this version takes away from the possibility?

w.r.t. the modern practices point, I understand what you mean, but personally I've seen far more people (sadly) fail to build out the basic logic effectively than become overwhelmed by seeing a class, heck we weren't even using any lifecycle hooks. 

Open to putting back in a class component somewhere if you think its worthwhile, but I've had more than a handful of candidates give this project a sideways look over the past 3 years. It's not exclusively an employers market, high quality candidates are evaluating what they see from us as well. Just my 2c.

**Comment by joepurnell1** [comment]
> I can happily remove the TODO comments to leave things more open ended

I think this would be good if other people think the same

> is there anything specific you have seen candidates do where this version takes away from the possibility?

Looking at it again with fresh eyes I think candidates will still see the same issues which should be ok. I thought we were just using enums not making the types from it. My bad

> I've seen far more people (sadly) fail to build out the basic logic effectively than become overwhelmed by seeing a class, heck we weren't even using any lifecycle hooks.

But isn't part of the work finding answers to your problems and collaborating with others? I encourage open book and making it a pairing session and these are things that highlight both cause we can open converse and find the answers together if they do struggle. If someone becomes overwhelmed by the use of a class I think thats a good sign of their level?
